common/models/battle-log-model.ts

```diff
@@ -1,321 +1,329 @@
 import {GameModel} from './game-model'
-
-export type BattleLogFormatT = Parameters<BattleLogModel['format']>
+import {LineNode, formatText} from '../utils/formatting'
+import {DEBUG_CONFIG} from '../config'
+import Card from '../cards/base/card'
+import {PickInfo} from '../types/server-requests'
+import StatusEffect from '../status-effects/status-effect'
 
 export class BattleLogModel {
 	private game: GameModel
-	private log: Array<BattleLogT>
+	private logMessageQueue: Array<BattleLogT>
 
 	constructor(game: GameModel) {
 		this.game = game
-		this.log = []
+
+		/** Log entries that still need to be processed */
+		this.logMessageQueue = []
+	}
+
+	private generateEffectEntryHeader(card: CardT | null): string {
+		const currentPlayer = this.game.currentPlayer.playerName
+		if (!card) return ''
+		const cardInfo = CARDS[card.cardId]
+
+		return `$p{You|${currentPlayer}}$ used $e${cardInfo.name}$ `
 	}
 
-	// BATTLE LOG STUFF //
-	private format(
-		text: string,
-		format: 'plain' | 'highlight' | 'player' | 'opponent',
-		condition: 'player' | 'opponent' | undefined = undefined
-	): MessageTextT {
-		const output: MessageTextT = {
-			text: text,
-			censoredText: text,
-			format: format,
+	private generateCoinFlipDescription(coinFlip: CurrentCoinFlipT): string {
+		const heads = coinFlip.tosses.filter((flip) => flip === 'heads').length
+		const tails = coinFlip.tosses.filter((flip) => flip === 'tails').length
+
+		if (coinFlip.tosses.length === 1) {
+			return heads > tails ? `flipped $gheads$` : `flipped $btails$`
+		} else if (tails === 0) {
+			return `flipped $g${heads} heads$`
+		} else if (heads === 0) {
+			return `flipped $b${tails} tails$`
+		} else {
+			return `flipped $g${heads} heads$ and $b${tails} tails$`
 		}
-		if (condition) output.condition = condition
-		return output
 	}
 
-	private sendBattleLogEntry() {
-		this.game.getPlayers().forEach((player) => {
-			player.socket.emit('BATTLE_LOG_ENTRY', {
-				type: 'BATTLE_LOG_ENTRY',
-				payload: this.log,
-			})
-		})
+	private generateCoinFlipMessage(
+		attack: AttackModel,
+		coinFlips: Array<CurrentCoinFlipT>
+	): string | null {
+		const entry = coinFlips.reduce((r: string | null, coinFlip) => {
+			const description = this.generateCoinFlipDescription(coinFlip)
+
+			if (coinFlip.opponentFlip) return r
+			if (HERMIT_CARDS[coinFlip.cardId] && attack.type === 'effect') return r
+			if (SINGLE_USE_CARDS[coinFlip.cardId] && attack.type !== 'effect') return r
 
-		while (this.log.length > 0) {
-			const lastEntry = this.log.pop()
-			if (!lastEntry) continue
+			return description
+		}, null)
+
+		return entry
+	}
+
+	public async sendLogs() {
+		while (this.logMessageQueue.length > 0) {
+			const firstEntry = this.logMessageQueue.shift()
+			if (!firstEntry) continue
 
 			this.game.chat.push({
 				createdAt: Date.now(),
-				message: lastEntry.description,
-				playerId: lastEntry.player,
+				message: formatText(firstEntry.description),
```

Comment:
Any player names in logs are not tested for profanity
```suggestion
				message: formatText(firstEntry.description, {censor: true}),
```